cli/internal/cloudcmd/upgrade.go

```diff
@@ -74,100 +76,150 @@ func NewUpgrader(outWriter io.Writer, log debugLog) (*Upgrader, error) {
 		stableInterface:  &stableClient{client: kubeClient},
 		dynamicInterface: &dynamicClient{client: unstructuredClient},
 		helmClient:       helmClient,
-		outWriter:        outWriter,
-		log:              log,
+
+		outWriter: outWriter,
+		log:       log,
 	}, nil
 }
 
-// UpgradeImage upgrades the cluster to the given measurements and image.
-func (u *Upgrader) UpgradeImage(ctx context.Context, newImageReference, newImageVersion string, newMeasurements measurements.M) error {
-	nodeVersion, err := u.getConstellationVersion(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieving current image: %w", err)
+// UpgradeHelmServices upgrade helm services.
+func (u *Upgrader) UpgradeHelmServices(ctx context.Context, config *config.Config, timeout time.Duration, allowDestructive bool) (string, error) {
+	err := u.helmClient.Upgrade(ctx, config, timeout, allowDestructive)
+	invalidUpgrade := &compatibility.InvalidUpgradeError{}
+	if errors.As(err, &invalidUpgrade) {
+		return fmt.Sprintf("Skipping Microservice upgrade: %s", err), nil
 	}
```

Comment:
Instead of this function parsing the error and returning a string in special cases, I would opt for letting the caller handle this.
Since the caller most likely has access to a writer, we would be able to directly print an error/warning instead of having to carry around multiple return values.

Alternatively, you could give `Upgrader` access to a writer and let this function handle the printing.